File: UserAuth.py
import sqlite3
import hashlib

# Connect to SQL database
conn = sqlite3.connect('UserAuth.db')
c = conn.cursor()

# Create users table
c. execute('''
     CREATE TABLE IF NOT EXISTS users (
         id INTEGER PRIMARY KEY,
         username TEXT UNIQUE,
         password TEXT
     )
''')
conn.commit()

def hash_password(password):
    return hashlib.sha256(password.encode()).hexdigest()

def register(username, password):
    hashed = hash_password(password)
    try:
        c.execute('INSERT INTO users (username, password) VALUES (?, ?)', (username, hashed))
        conn.commit()
        print("User registered successfully")
    except sqlite3.IntegrityError:
        print("Username already exists")

def login(username, password):
    hashed = hash_password(password)
    c.execute('SELECT password FROM users WHERE username = ?', (username,))
    result = c.fetchone()
    if result and result[0] == hashed:
        print("Login successful!")
    else:
        print("Invalid username or password.")

def display_menu():
    print("Welcome to the User Authenticator")
    print("1. Register")
    print("2. Login")
    print("3. Exit")

def main():
    while True:
        display_menu()
        choice = input("Please choose an option (1-3): ")

        if choice == '1':
            username = input("Enter a username: ")
            password = input("Enter a password: ")
            register(username, password)
        elif choice == '2':
            username = input("Enter your username: ")
            password = input("Enter your password: ")
            login(username, password)
        elif choice == '3':
            print("Exiting the program.")
            break
        else:
            print("Invalid option. Please try again.")

if __name__ == "__main__":
    main()

# Example usage
register("testuser", "test123")
login("testuser", "test123")

conn.close()

File: UserAuth.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <iomanip>

// Hold user data
struct User {
    std::string username;
    std::string password;
};

// Read user data from file
std::vector<User> readUserDataFromFile(const std::string& filename) {
    std::vector<User> users;
    std::ifstream file(filename);

    if (file.is_open()) {
        std::string username, password;

        while (file >> username >> password) {
            users.push_back({username, password});
        }

        file.close();
    }

    return users;
}

// Write user data to file
void writeUserDataToFile(const std::string& filename, const std::vector<User>& users) {
    std::ofstream file(filename);

    if (file.is_open()) {
        for (const User& user : users) {
            file << user.username << " " << user.password << "\n";
        }

        file.close();
    }
}

// Check is username exists
bool doesUsernameExist(const std::vector<User>& users, const std::string& username) {
    for (const User& user : users) {
        if (user.username == username) {
            return true;
        }
    }

    return false;
}

// Register new user
void registerUser(std::vector<User>& users) {
    std::string username, password;

    std::cout << "\n=== Register ===\n";
    std::cout << "Enter a username: ";
    std::cin >> username;

    if (doesUsernameExist(users, username)) {
        std::cout << "Username already exists. Please choose a different username.\n";

        return;
    }

    std::cout << "Enter a password: ";
    std::cin >> password;

    users.push_back({username, password});
    writeUserDataToFile("users.txt", users);

    std::cout << "User registered successfully.\n";
}

// Authenticate User
bool authenticateUser(const std::vector<User>& users, const std::string& username, const std::string& password) {
    for (const User& user : users) {
        if (user.username == username && user.password == password) {
            return true;
        }
    }

    return false;
}

// Header
void displayHeader(const std::string& title) {
    std::cout << "\033[1;36m"; // cyan text
    std::cout << std::setw(40) << std::right << title << std::endl;
    std::cout << "\033[0m"; // Reset text color
}

int main() {
    std::vector<User> users = readUserDataFromFile("users.txt");

    int choice;
    std::string username, password;

    do {
        std::cout << "\n=== Main Menu ===\n";
        std::cout << "1. Register\n";
        std::cout << "2. Login\n";
        std::cout << "3. Exit\n";
        std::cout << "Enter your choice: ";
        std::cin >> choice;

        switch (choice) {
            case 1:
                registerUser(users);
                break;
            case 2:
                std::cout << "\n=== Login ===\n";
                std::cout << "Enter username: ";
                std::cin >> username;
                std::cout << "Enter password: ";
                std::cin >> password;

                if (authenticateUser(users, username, password)) {
                    displayHeader("Login successful!");
                } else {
                    displayHeader("Invalid username or password.");
                }
                break;
            case 3:
                displayHeader("Exiting...");
                break;
            default:
                displayHeader("Invalid choice. Please try again.");
        }
    } while (choice != 3);

    return 0;
}